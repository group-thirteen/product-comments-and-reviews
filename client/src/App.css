
.AppContainer {
  padding: 24px;
  border-top: 1px solid rgba(0, 0, 0, .7);
}

