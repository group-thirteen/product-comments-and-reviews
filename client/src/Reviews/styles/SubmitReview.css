
.myModal {
  position: absolute;
  top: 200px;
  left: 200px;
  right: 200px;
  bottom: 350px;
}

.myOverlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: whitesmoke;
}

.label {
  display: inline-block;
  border: 1px solid black;
  margin: 0;
  width: 80%;
  height: 36px;
  background: white;
  text-align: center;
  color: black;
  padding: 24px;
  margin: 18px 0;
}

.submitInput {
  width: 100%;
  height: 36px;
  
}

.reviewSubmit {
  display: inline;
}