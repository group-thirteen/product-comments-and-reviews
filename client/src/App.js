import React from 'react'
import { ajax } from 'jquery'
import Title from './Title'
import ReviewList from './Reviews/Components/ReviewList.js'
import Info from './Reviews/Components/ReviewSummary.js'
import FilterContainer from './Reviews/Components/FilterContainer'
import styles from './App.css'
import './global.css'
import ReactPaginate from 'react-paginate';

class Reviews extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      reviews: [],
      average: 0,
      stars: {},
      filteredReviews: null,
      showForm: false,
      pagination: {
        offset: 0,
        perPage: 5
      }

    }
    this.get = this.get.bind(this)
    this.handleNewReviews = this.handleNewReviews.bind(this)
    this.getRatingTotal = this.getRatingTotal.bind(this)
    this.post = this.post.bind(this)
    this.handleRatingSelection = this.handleRatingSelection.bind(this)
    this.sortStarRating = this.sortStarRating.bind(this)
    this.sortByTime = this.sortByTime.bind(this)
  }

  componentDidMount () {
    this.get()
  }

  get () {
    ajax({
      method: 'GET',
      url: '/reviews',
      data: {},
      success: (reviews) => { this.handleNewReviews(reviews) },
      error: () => { console.log('Failed in the get') }
    })
  }
  post (review) {
    console.log('review before', review)
    ajax({
      method: 'POST',
      url: '/reviews',
      data: review,
      success: (reviews) => { this.handleNewReviews(reviews) },
      error: () => {console.log('error in the post')}
    })  
  }

  handleNewReviews (reviews) {
    let sum = 0
    reviews.forEach((review) => {
      sum += review.score
    })
    const average = sum / reviews.length
    this.getRatingTotal(reviews)
    this.setState({ average, reviews, pageCount: Math.ceil(reviews.length / this.state.pagination.perPage) }, () => (console.log('this.state' ,this.state)))
    this.sortByTime()

  }

  getRatingTotal (reviews) {
    let stars = {
      1: 0,
      2: 0, 
      3: 0,
      4: 0,
      5: 0
    }
    reviews.forEach((review) => {
      stars[review.score]++
    })
    this.setState({ stars })
  }



  handleRatingSelection (rating) {
    const filtered = this.state.reviews.filter((review) => (
      (rating === 0) || (review.score === rating)
      ))
    this.setState({ filteredReviews: filtered })
  }
  
  sortStarRating() {
    const sorted = this.state.reviews.sort((a, b) => {
      if (a.score > b.score) {
        return -1
      }
      if (a.score < b.score) {
        return 1
      }
      return 0
    })
    this.setState({ reviews: sorted })
  }

  sortByTime () {
    const sorted = this.state.reviews.sort((a, b) => {
      if (a.date > b.date) {
        return -1
      }
      if (a.date < b.date) {
        return 1
      }
      return 0
    })
    this.setState({ reviews: sorted })
  }

  handlePageClick () {
    const getPage = (offSet, reviewsPerPage) => {
      
    }
  }  


  render () {
    return (
      <div className={styles.AppContainer}>
        <Title />
        <Info 
          average={ this.state.average } 
          post={ this.post } 
          showForm={ this.state.showForm } 
          onButtonClick={() => (this.setState({showForm: true}))} 
          stars={this.state.stars} 
          reviewTotal={this.state.reviews.length} 
          />
        <FilterContainer 
          handleRatingSelection={this.handleRatingSelection} 
          starSortFunc={this.sortStarRating} 
          timeSortFunc={this.sortByTime} />
        <ReviewList reviews={ this.state.filteredReviews || this.state.reviews } />
        <ReactPaginate
          previousLabel={'previous'}
          nextLabel={'next'}
          breakLabel={'...'}
          breakClassName={'break-me'}
          pageCount={5}
          marginPagesDisplayed={1}
          pageRangeDisplayed={2}
          onPageChange={this.handlePageClick}
          containerClassName={'pagination'}
          subContainerClassName={'pages pagination'}
          activeClassName={'active'}
        />
      </div>
    )
  }
}

export default Reviews
