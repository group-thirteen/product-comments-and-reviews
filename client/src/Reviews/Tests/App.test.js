
import React from 'react'
import { shallow, mount } from 'enzyme'
import App from '../../App.js'
import { createSerializer } from 'enzyme-to-json'
import styles from '../Components/components.css'
// console.log('STYLES',styles())

jest.mock('jquery', () => ({ ajax: jest.fn((options) => {

  options.success(
    [
      {  title: 'title1',
        username: 'username1',
        body: 'body1',
        score: 1,
        date: 10
      },
      {  title: 'title2',
          username: '2',
          body: 'body2',
          score: 5,
          date: 20
      },
      {  title: 'title3',
          username: 'username3',
          body: 'body3',
          score: 3,
          date: 30
      }
    ]
  )   
}) }))
jest.mock('../Components/components.css', () => jest.fn())
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }))

describe('App', () => {
  it('should render and have correct state', () => {
    const wrapped = shallow(<App />)
    expect(wrapped).toMatchSnapshot()
    expect(wrapped.state().stars).toEqual( { '1': 1, '2': 0, '3': 1, '4': 0, '5': 1 })
    expect(wrapped.state().average).toEqual(3)
    expect(wrapped.state().reviews.length).toEqual(3)
  })
  it('should filter reviews', () => {
    const wrapper = mount(<App />)
    const form  = wrapper.find('.filter-form')
    const select = form.find('select')
    select.simulate('change', {target: {value: 2}})
    
    expect(wrapper.state().filteredReviews.length).toEqual(0)
    select.simulate('change', {target: {value: 1}})
    expect(wrapper.state().filteredReviews.length).toEqual(1)
  })
  it('should sort review by star or date', () => {
    const wrapper = mount(<App />)
    expect(wrapper).toMatchSnapshot()
    // const form  = wrapper.find('.sort-form')
    const select = wrapper.find('select')
    select.simulate('change', {target: {value: 'stars'}})
    select.simulate('submit', {})
    expect(wrapper.state().reviews[2].title).toEqual('title2')
    select.simulate('change', {target: {value: 'time'}})
    select.simulate('submit', {})
    expect(wrapper.state().reviews[2].title).toEqual('title3')
  })
  // it('should', () => {})

})
