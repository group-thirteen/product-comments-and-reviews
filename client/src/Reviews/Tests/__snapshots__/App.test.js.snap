// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render and have correct state 1`] = `
<div>
  <Title />
  <ReviewSummary
    average={3}
    onButtonClick={[Function]}
    post={[Function]}
    reviewTotal={3}
    showForm={false}
    stars={
      Object {
        "1": 1,
        "2": 0,
        "3": 1,
        "4": 0,
        "5": 1,
      }
    }
  />
  <FilterContainer
    handleRatingSelection={[Function]}
    starSortFunc={[Function]}
    timeSortFunc={[Function]}
  />
  <ReviewList
    reviews={
      Array [
        Object {
          "body": "body3",
          "date": 30,
          "score": 3,
          "title": "title3",
          "username": "username3",
        },
        Object {
          "body": "body2",
          "date": 20,
          "score": 5,
          "title": "title2",
          "username": "2",
        },
        Object {
          "body": "body1",
          "date": 10,
          "score": 1,
          "title": "title1",
          "username": "username1",
        },
      ]
    }
  />
</div>
`;

exports[`App should sort review by star or date 1`] = `
<div>
  <div>
    <h1>
      Reviews
    </h1>
  </div>
  <div>
    <div>
      <div>
        <div
          className="star-ratings"
          style={
            Object {
              "boxSizing": "border-box",
              "display": "inline-block",
              "position": "relative",
            }
          }
          title="3 Stars"
        >
          <svg
            className="star-grad"
            style={
              Object {
                "height": "0",
                "position": "absolute",
                "visibility": "hidden",
                "width": "0",
                "zIndex": "0",
              }
            }
          >
            <defs>
              <linearGradient
                id="starGrad147098523954110"
                x1="0%"
                x2="100%"
                y1="0%"
                y2="0%"
              >
                <stop
                  className="stop-color-first"
                  offset="0%"
                  style={
                    Object {
                      "stopColor": "black",
                      "stopOpacity": "1",
                    }
                  }
                />
                <stop
                  className="stop-color-first"
                  offset="0%"
                  style={
                    Object {
                      "stopColor": "black",
                      "stopOpacity": "1",
                    }
                  }
                />
                <stop
                  className="stop-color-final"
                  offset="0%"
                  style={
                    Object {
                      "stopColor": "rgb(203, 211, 227)",
                      "stopOpacity": "1",
                    }
                  }
                />
                <stop
                  className="stop-color-final"
                  offset="100%"
                  style={
                    Object {
                      "stopColor": "rgb(203, 211, 227)",
                      "stopOpacity": "1",
                    }
                  }
                />
              </linearGradient>
            </defs>
          </svg>
          <div
            className="star-container"
            onClick={null}
            onMouseEnter={null}
            onMouseLeave={null}
            style={
              Object {
                "cursor": undefined,
                "display": "inline-block",
                "paddingLeft": undefined,
                "paddingRight": "0px",
                "position": "relative",
                "verticalAlign": "middle",
              }
            }
          >
            <svg
              className="widget-svg"
              style={
                Object {
                  "height": "24px",
                  "transform": undefined,
                  "transition": "transform .2s ease-in-out",
                  "width": "24px",
                }
              }
              viewBox="0 0 51 48"
            >
              <path
                className="star"
                d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                style={
                  Object {
                    "fill": "black",
                    "transition": "fill .2s ease-in-out",
                  }
                }
              />
            </svg>
          </div>
          <div
            className="star-container"
            onClick={null}
            onMouseEnter={null}
            onMouseLeave={null}
            style={
              Object {
                "cursor": undefined,
                "display": "inline-block",
                "paddingLeft": "0px",
                "paddingRight": "0px",
                "position": "relative",
                "verticalAlign": "middle",
              }
            }
          >
            <svg
              className="widget-svg"
              style={
                Object {
                  "height": "24px",
                  "transform": undefined,
                  "transition": "transform .2s ease-in-out",
                  "width": "24px",
                }
              }
              viewBox="0 0 51 48"
            >
              <path
                className="star"
                d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                style={
                  Object {
                    "fill": "black",
                    "transition": "fill .2s ease-in-out",
                  }
                }
              />
            </svg>
          </div>
          <div
            className="star-container"
            onClick={null}
            onMouseEnter={null}
            onMouseLeave={null}
            style={
              Object {
                "cursor": undefined,
                "display": "inline-block",
                "paddingLeft": "0px",
                "paddingRight": "0px",
                "position": "relative",
                "verticalAlign": "middle",
              }
            }
          >
            <svg
              className="widget-svg"
              style={
                Object {
                  "height": "24px",
                  "transform": undefined,
                  "transition": "transform .2s ease-in-out",
                  "width": "24px",
                }
              }
              viewBox="0 0 51 48"
            >
              <path
                className="star"
                d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                style={
                  Object {
                    "fill": "black",
                    "transition": "fill .2s ease-in-out",
                  }
                }
              />
            </svg>
          </div>
          <div
            className="star-container"
            onClick={null}
            onMouseEnter={null}
            onMouseLeave={null}
            style={
              Object {
                "cursor": undefined,
                "display": "inline-block",
                "paddingLeft": "0px",
                "paddingRight": "0px",
                "position": "relative",
                "verticalAlign": "middle",
              }
            }
          >
            <svg
              className="widget-svg"
              style={
                Object {
                  "height": "24px",
                  "transform": undefined,
                  "transition": "transform .2s ease-in-out",
                  "width": "24px",
                }
              }
              viewBox="0 0 51 48"
            >
              <path
                className="star"
                d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                style={
                  Object {
                    "fill": "rgb(203, 211, 227)",
                    "transition": "fill .2s ease-in-out",
                  }
                }
              />
            </svg>
          </div>
          <div
            className="star-container"
            onClick={null}
            onMouseEnter={null}
            onMouseLeave={null}
            style={
              Object {
                "cursor": undefined,
                "display": "inline-block",
                "paddingLeft": "0px",
                "paddingRight": undefined,
                "position": "relative",
                "verticalAlign": "middle",
              }
            }
          >
            <svg
              className="widget-svg"
              style={
                Object {
                  "height": "24px",
                  "transform": undefined,
                  "transition": "transform .2s ease-in-out",
                  "width": "24px",
                }
              }
              viewBox="0 0 51 48"
            >
              <path
                className="star"
                d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                style={
                  Object {
                    "fill": "rgb(203, 211, 227)",
                    "transition": "fill .2s ease-in-out",
                  }
                }
              />
            </svg>
          </div>
        </div>
        (
        0
        )
      </div>
    </div>
    <div>
      <div
        className="star-overview"
      >
        <div>
          <span>
            5
             
            stars
          </span>
          <div>
            <div
              style={
                Object {
                  "width": "Infinity%",
                }
              }
            />
          </div>
          <span>
            1
          </span>
        </div>
        <div>
          <span>
            4
             
            stars
          </span>
          <div>
            <div
              style={
                Object {
                  "width": "NaN%",
                }
              }
            />
          </div>
          <span>
            0
          </span>
        </div>
        <div>
          <span>
            3
             
            stars
          </span>
          <div>
            <div
              style={
                Object {
                  "width": "Infinity%",
                }
              }
            />
          </div>
          <span>
            1
          </span>
        </div>
        <div>
          <span>
            2
             
            stars
          </span>
          <div>
            <div
              style={
                Object {
                  "width": "NaN%",
                }
              }
            />
          </div>
          <span>
            0
          </span>
        </div>
        <div>
          <span>
            1
             
            star  
          </span>
          <div>
            <div
              style={
                Object {
                  "width": "Infinity%",
                }
              }
            />
          </div>
          <span>
            1
          </span>
        </div>
      </div>
    </div>
    <div>
      <div>
        <button
          onClick={[Function]}
        >
          Write a Review
        </button>
      </div>
    </div>
  </div>
  <div>
    <div>
      <form>
        <label>
          <select
            onChange={[Function]}
            value="sort"
          >
            <option
              value="all"
            >
              Sort reviews
            </option>
            <option
              value="stars"
            >
              Star Rating
            </option>
            <option
              value="time"
            >
              Submission Time
            </option>
          </select>
        </label>
      </form>
    </div>
    <div>
      <form
        className="filter-form"
      >
        <label>
          <select
            onChange={[Function]}
          >
            <option
              value={0}
            >
              star rating
            </option>
            <option
              value={1}
            >
              One
            </option>
            <option
              value={2}
            >
              Two
            </option>
            <option
              value={3}
            >
              Three
            </option>
            <option
              value={4}
            >
              Four
            </option>
            <option
              value={5}
            >
              Five
            </option>
          </select>
        </label>
      </form>
    </div>
  </div>
  <div />
</div>
`;
